### ISSUE TITLE
Crash on Arduino101 if fiber is CPU hog

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#329]
**_Reported by Szymon Janc:_**

If fiber is hogging CPU (even with simple loop) crash occurs. This happens on Arduino 101 but doesn't happen on qemu target (maybe simply due to too fast cpu?).

***** CPU exception 6
Current thread ID = 0xa8006d80
Faulting segment:address = 0x00000008:0xa8006d24
eax: 0x00000000, ebx: 0x400303f6, ecx: 0x00006d01, edx: 0xa8007d44
esi: 0x00000000, edi: 0x00000000, ebp: 0a8007d48, esp: 0xa8008cb0
eflags: 0x00010246
Fatal ISR error! Spinning...

Simple test app that shows the problem is available at https://gerrit.zephyrproject.org/r/#/c/2043/

With this app doing 3,5M iterations is OK, but 4M or more is causing crash.

(Imported from Jira ZEP-370)

**Comment by nashif** [comment]
**_by Sharron LIU:_**

I tried the "tc_crash" on my board, and I found if I comment out below lines, the fiber loop can be done successfully.
1. comment out CONFIG_INIT_STACKS=y from "prj.conf"
2. comment out stack_analyze() called from "main.c"

May this suggest further investigation on stack related?

**Comment by nashif** [comment]
**_by Peter Mitsis:_**

For what it is worth, I have been able to reproduce this on QEMU by bumping up the number of iterations in the test app to 500 million.

My output is ...

start
result 1711656320 
done
fiber start
***** CPU exception 7
Current thread ID = 0x0010350c
Faulting segment:address = 0x00000008:0x001034f9
eax: 0x00103438, ebx: 0xaaaaaaaa, ecx: 0x00104518, edx: 0x001044bc
esi: 0xaaaaaaaa, edi: 0xaaaaaaaa, ebp: 0001044e0, esp: 0x00104d50
eflags: 0x00000203
Fatal ISR error! Spinning...

1. CPU exception 7: corresponds to Device not available (No math coprocessor).
2. It was trying to execute at address 0x1034f9, which is beyond the text section.
3. If interrupts are locked during the test app's tc_test() routine, I do not experience a crash on QEMU.
4. From this, I take it that some sort of corruption or error is occurring while processing interrupts. This does not mean that the problem is still not stack related.

**Comment by nashif** [comment]
**_by Peter Mitsis:_**

In a microkernel, when a fiber runs for too long without yielding or blocking the tick events pile up in the kernel server's command stack.  The kernel server's command stack is only emptied by the kernel server fiber, which is not getting a chance to run since the running fiber has neither yielded nor blocked.

Result: The kernel server's command stack overflows due to unprocessed tick events thereby corrupting memory and the system crashes.

Option 1:
Increase the size of the kernel server command stack size through CONFIG_COMMAND_STACK_SIZE (default 64). This is expected to be brittle and very dependent upon the speed of the platform/board.

Option 2:
Use a task instead of fiber for the processing since that can be preempted to allow the microkernel tick to be processed.

Option 3:
Redo the fiber's work flow to either yield or block periodically so that the kernel server fiber has a chance to run.

**Comment by nashif** [comment]
**_by Szymon Janc:_**

OK, we already have GH-330 to offloading Bluetooth ECC calculations to task so I'll go with option 2 for BT code.

But how we should proceed with nanokernel configuration? Is fiber allowed to block CPU for long there?

**Comment by nashif** [comment]
**_by Johan Hedberg:_**

As Szymon mentioned this issue comes with the TinyCrypt APIs for ECC where public key generation or shared secret calculation easily takes a couple of seconds with Quark SE. Option 1 is indeed brittle, and option 3 would require substantial changes to the TinyCrypt implementation and/or API. So that leaves us with option 2, i.e. we'll need to create a Bluetooth stack-specific task for ECC related operations.

I can easily imagine other people hitting this issue with different use cases however. To ease debugging, would it be possible to improve the resulting error of the command stack overflowing? Both I and Szymon ended up spending several days trying to understand what was going on (e.g. initially suspecting a bug in the TinyCrypt implementation) before realizing that this was something more generic related to fiber operation.

**Comment by nashif** [comment]
**_by Peter Mitsis:_**

In a nanokernel system, a greedy fiber will not impact the delivery of system ticks as those are done at interrupt level.

However, that is not to say that it might not have an impact on other drivers or devices in the system as they may defer some processing to a fiber and whatever queues they use (if any) could also overflow if a fiber (or a set of fibers) becomes too greedy.

In that light, the best suggestion that I can currently offer is to try and limit the amount of continuous work that a fiber must do so that there are opportunities for it to block and/or yield to let another fiber execute (if one is available).

**Comment by nashif** [comment]
**_by Johan Hedberg:_**

As I said, doing intermediate yields from the fiber would require changing the TinyCrypt API and implementation. Currently all the time ends up being spent in a single public TinyCrypt API call. I'll leave it for Connie to comment on the feasibility of fixing that.

The Bluetooth subsystem fibers are agnostic to whether they run in a nano or micro kernel, so the only short term solution is to add a microkernel dependency whenever the use of TinyCrypt ECC has been chosen in the configuration and then use a dedicated task to offload these calculations.

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

Not a bug: improper use of fibers in a microkernel. Fibers should be used for short un-interruptible bursts of work only. If they have to do long amounts of work, they should yield to higher-priority fibers.

**Comment by nashif** [comment]
**_by Luiz Augusto von Dentz:_**

It is understandable that we may not consider this a bug, but detecting should probably be made easier or at least document, or have a Kconfig, saying how many ticks the fiber could take otherwise I fear we will make people lose a lot of time trying to figure out what is going on just to reach the same conclusion.

**Comment by nashif** [comment]
**_by Johan Hedberg:_**

I agree with Luiz here. There should be a clear indication of this happening instead of just arbitrary memory starting to be overwritten. Not even with CONFIG_DEBUG enabled did we get any better indications of what was wrong. What makes this worse is that this happens with a public API that's not in any way documented to be forbidden to be called from a fiber in a microkernel configuration. So I hope we could at least agree on a couple of action points:

 1. Make it easier to discover/debug this. It's very unlikely that we'll be the only/last ones to hit this issue.
 2. Document any public API that may take a considerable amount of CPU time without yielding to not be usable from a microkernel fiber

**Comment by nashif** [comment]
**_by Luiz Augusto von Dentz:_**

Actually after looking what the code actually do it does seems it is not really related to time but to how many commands can be pushed while a fiber run, so each platform will have different limits on how long a fiber can run to the extend that fiber becomes very similar to an interrupt doing as little work as possible not to hit that limit.

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

assert on k_server command stack overflow: https://gerrit.zephyrproject.org/r/#/c/2085/

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

The root cause is indeed the number of command packets that can be pushed on the stack. However, the subsystem that will most likely overflow the stack is the timer driver that pushes an event on every kernel tick, so it is time-based as well.

**Comment by nashif** [comment]
**_by Johan Hedberg:_**

Would there be any sense in trying to reduce the impact on the stack usage from tick events by creating a solution where there is ever only one such event on the stack? This event would simply say "ticks changed" and when receiving it the kernel server would check the current tick count, deduce the number of ticks since the last time and act accordingly. There'd also need to be some external variable to track that there's a "ticks changed" event on the stack so that no more than one is pushed there. Or would such a solution cause too much overhead and complexity?

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

It's possible that this would work, but it has to be investigated carefully, especially because of tickless idle. On some platforms, at least the ones using LOAPIC IIRC, the tickless idle relies on the ticks being posted as events to the microkernel.

The point still stands that a fiber, in a microkernel, should not be a CPU hog.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Hi _Benjamin Walsh_ .  What are the appropriate next steps for this bug?  It's been sitting in the Resolved state for a while and I'd like to get it moving along as appropriate.  Also, is the "Won't Fix" resolution still valid?  If not, please clear it.  Thx.
I see Anas tagged it for v1.4.0, but that release is done so I'm wondering if it needs to be fixed in v1.5.0 now.  I'm cc'ing _Inaky Perez-Gonzalez_  since he's covering for Anas.
_Sharron LIU_

**Comment by nashif** [comment]
**_by Sharron LIU:_**

Verified assertion takes effect in the tc_crash testapp.
```
# source zephyr-env.sh
# git fetch https://gerrit.zephyrproject.org/r/zephyr refs/changes/43/2043/3 && git format-patch -1 --stdout FETCH_HEAD > tc_crash.patch
# git apply tc_crash.patch
# echo CONFIG_ASSERT=y >>samples/bluetooth/tc_crash/prj.conf
# echo CONFIG_ASSERT_LEVEL=2 >>samples/bluetooth/tc_crash/prj.conf
# make -C samples/bluetooth/tc_crash BOARD=quark_se_devboard flash
```

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

Won't fix is still the correct resolution, especially with the unified kernel coming down the pipe, where command packets will be a thing of the past. It was not a bug either, but working as intended, since this was due to a user error, that did not yield periodically to a fiber of equal or higher priority. The assert statement is the correct way of handling the issue in the current kernel.

**Comment by nashif** [comment]
**_by Benjamin Walsh:_**

I can put it in a 'fixed' state if you prefer, since that assert statement is the proper "fix" (even if I don't really call it a fix, nor was the issue a real bug).

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Reopening so it can be confirmed and closed by reporter.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Szymon Janc_ , please review the comments on this bug and if you're okay with what's been done, please close it.  If you're not okay, please explain why.  Thanks.

**Comment by nashif** [comment]
**_by Szymon Janc:_**

As noted by Ben this was due to invalid use of fiber. The code that was root cause of this report is also fixed now (we use task on microkernel for Bluetooth ECC calculation) so this can be closed.

**Comment by nashif** [comment]
**_by Sharron LIU:_**

_Szymon Janc_  Please could you change the bug status to verified than closed, per you comment above? Thanks.

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

_Szymon Janc_ , please see _Sharron LIU_ 's comment above.  Also, as the Reporter of this bug we need you to verify it's fixed since it's in the Resolved state.  Please verify it ASAP so we know whether or not more work is required to fix it before the upcoming 1.5 release.  If it's fixed, please move it to the Verified state.  Thanks!

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Correcting the priority field
